src/Streamly/Internal/Network/Socket.hs

```diff
@@ -467,13 +469,20 @@ read = UF.first defaultChunkSize readWith
 -- Writing
 -------------------------------------------------------------------------------
 
+-- XXX We should not define it here. This module should not decide whether we
+-- use D or K. Maintanince will become tiresome. We should have it as
+-- Stream.Eliminate.mapM_
+{-# INLINE mapM'_ #-}
+mapM'_ :: Monad m => (a -> m b) -> S.Stream m a -> m ()
+mapM'_ f = D.mapM_ f . S.toStreamD
+
 -- | Write a stream of arrays to a handle.
 --
 -- @since 0.7.0
 {-# INLINE putChunks #-}
 putChunks :: (MonadIO m, Unboxed a)
     => Socket -> Stream m (Array a) -> m ()
-putChunks h = S.mapM_ (liftIO . writeChunk h)
+putChunks h = mapM'_ (liftIO . writeChunk h)
```

Comment:
You can use `S.fold FL.drain $ S.mapM $ ...` and check the benchmarks.

Reply:
Use `Fold.drainBy`.